flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -1109,6 +1127,8 @@ public void releaseOrPromotePartitions(
             offerSlotsToJobManager(jobId);
         }
 
+        tryPersistAllocationSnapshot(slotId, jobId, targetAddress, allocationId, resourceProfile);
```

Comment:
I think we should persist the information before offering the slots.

Reply:
Why? Couldn't it cause some inconsistencies? Given the recovery is best effort, wouldn't be better to end up in a state where the slot is offered, but no information about it is persisted, more than the contrary?

Reply:
I think it is a good pattern to do state changes and then communicate with external systems. In this case, it might not matter a lot but in the general case it might happen that the external system learns about a state that you cannot recover if you fail after talking to the other system but before persisting it. If we fail after persisting but before communicating with an external system, then we can simply retry it in case of a recovery.

Reply:
> In this case, it might not matter a lot but in the general case it might happen that the external system learns about a state that you cannot recover if you fail after talking to the other system but before persisting it. If we fail after persisting but before communicating with an external system, then we can simply retry it in case of a recovery.

Agree, but in this case, if we fail to persist the allocation snapshot, nothing will break, because it's best-effort. Isn't it better than ending up with the state persisted, but without notifying to the external system? What happens if in the meantime, the TM fails? Isn't the restore of the TM going to signal to the JM that it restored a specific slot allocation, which in turn the JM is not expecting because it never received the ack for such allocation?